File: genomevo/modules/bactcg_src/postprocess.py
import argparse
import logging
import os
import shutil
from concurrent.futures import ThreadPoolExecutor
import subprocess

logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")

def step2_get_fa(cdhit_dir, cg_file, out_dir):
    logging.info("Starting Step 2: get_fa_file-all")
    os.makedirs(out_dir, exist_ok=True)
    seq = {}
    for filename in os.listdir(cdhit_dir):
        if not filename.endswith('.fasta'): continue
        indexName = filename.replace('.fasta', '')
        with open(os.path.join(cdhit_dir, filename), 'r') as f:
            current_seq = ""
            for line in f:
                if line.startswith('>'):
                    current_seq = ">" + indexName + "---" + line.strip()[1:]
                    seq[current_seq] = ""
                elif line.strip():
                    seq[current_seq] += line.strip()
    
    with open(cg_file, 'r') as f:
        headers = f.readline().strip().split('\t')
        indexList = [h.strip() + "---" for h in headers]
        
        for line in f:
            if not line.strip(): continue
            parts = line.strip().split('\t')
            outLine = [indexList[i] + parts[i].strip() for i in range(min(len(parts), len(indexList)))]
            
            if not outLine: continue
            out_filename = outLine[0].replace(indexList[0], '')
            
            with open(os.path.join(out_dir, out_filename + ".fa"), 'w') as outf:
                for el in outLine:
                    k = ">" + el
                    if k in seq:
                        outf.write(k + "\n" + seq[k] + "\n")

def step3_get_mega(fa_dir, out_meg_dir):
    logging.info("Starting Step 3: get_mega-all")
    os.makedirs(out_meg_dir, exist_ok=True)
    fa_files = [f for f in os.listdir(fa_dir) if f.endswith('.fa')]
    
    # Try resolving clustalw2 from $PATH, or use local
    clustalw2 = shutil.which("clustalw2")
    if not clustalw2:
        clustalw2 = os.path.abspath("Tool/clustalw2")
    
    gcg2meg = os.path.abspath("Tool/gcg2meg.pl")
        
    def run_clustal(f):
        base = f.replace('.fa', '')
        in_path = os.path.abspath(os.path.join(fa_dir, f))
        out_gcg = os.path.abspath(os.path.join(out_meg_dir, base + '.gcg'))
        out_meg = os.path.abspath(os.path.join(out_meg_dir, base + '.meg'))
        
        # Run Clustal
        subprocess.run([clustalw2, f"-infile={in_path}", "-type=protein", "-output=gcg", f"-outfile={out_gcg}", "-pwmatrix=GONNET", "-pwgapopen=10", "-pwgapext=0.1", "-gapopen=10", "-gapext=0.2", "-gapdist=4", "-align", "-quiet"])
        
        # Convert GCG to MEG using the native perl script as the original did
        if os.path.exists(out_gcg):
            subprocess.run(["perl", gcg2meg, out_gcg, out_meg])

    with ThreadPoolExecutor(max_workers=os.cpu_count()) as executor:
        executor.map(run_clustal, fa_files)

def step4_get_snp_mega(meg_dir, final_out, log_out):
    logging.info("Starting Step 4: get_SNP_mega-all with indel reduction pipeline")
    
    # 1. Process indels for each .meg mapping
    reduced_dir = os.path.join(os.path.dirname(meg_dir), "4.SNP_mega_temp")
    os.makedirs(reduced_dir, exist_ok=True)
    
    for f in os.listdir(meg_dir):
        if not f.endswith(".meg"): continue
        file_path = os.path.join(meg_dir, f)
        
        prot_dict = {}
        with open(file_path, 'r') as p1:
            faa_name = None
            for i, line in enumerate(p1):
                if i < 4: continue
                line = line.strip()
                if line.startswith("#") and not line.startswith("#MEGA"):
                    faa_name = line
                    prot_dict[faa_name] = ""
                elif faa_name:
                    prot_dict[faa_name] += line
                    
        if not prot_dict:
            continue

        line_one = list(prot_dict.values())[0]
        
        del_num_list = set()
        for col_idx in range(len(line_one)):
            ref_aa = line_one[col_idx]
            identical = True
            for seq in prot_dict.values():
                if len(seq) > col_idx and seq[col_idx] != ref_aa:
                    identical = False
                    break
            if identical:
                del_num_list.add(col_idx)

        prot_dict_new = {}
        for key, seq in prot_dict.items():
            new_seq = "".join([seq[i] for i in range(len(seq)) if i not in del_num_list])
            prot_dict_new[key] = new_seq

        with open(os.path.join(reduced_dir, "new-" + f), "w", encoding="utf-8") as outf:
            outf.write("#mega\n!Title CG;\n!Format DataType=Protein indel=-;\n\n")
            for k, v in prot_dict_new.items():
                outf.write(k + "\n" + v + "\n\n")

    # 2. Combine reduced megas
    prot_sequence_dict = {}
    prot_log_dict = {}
    
    for i, f in enumerate(os.listdir(reduced_dir)):
        file_path = os.path.join(reduced_dir, f)
        
        with open(file_path, 'r') as p1:
            faa_name = None
            for line_idx, line in enumerate(p1):
                if line_idx < 4: continue
                line = line.strip()
                if not line: continue
                if line.startswith("#") and not line.startswith("#mega"):
                    # Original logic strips down to strain name
                    faa_name = line.replace("---", "").replace("#", "").split()[0]
                    if i == 0:
                        prot_sequence_dict[faa_name] = ""
                elif faa_name:
                    if i == 0:
                        prot_sequence_dict[faa_name] += line
                        prot_log_dict[f] = len(prot_sequence_dict[faa_name])
                    else:
                        prev_len = len(prot_sequence_dict.get(faa_name, ""))
                        if faa_name in prot_sequence_dict:
                            prot_sequence_dict[faa_name] += line
                            prot_log_dict[f] = len(prot_sequence_dict[faa_name]) - prev_len
                            
    os.makedirs(os.path.dirname(final_out), exist_ok=True)
    with open(final_out, "w", encoding="utf-8") as output_line:
        output_line.write("#mega\n!Title CG;\n!Format DataType=Protein indel=-;\n\n")
        for k, v in prot_sequence_dict.items():
            output_line.write("#" + k + "\n" + v + "\n\n")

    with open(log_out, "w", encoding="utf-8") as log_file:
        for k, v in prot_log_dict.items():
            log_file.write(f"{k}\t{v}\n")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Post-process CG results")
    parser.add_argument("--cdhit-dir", default="output/CG_results/1.cd-hit_output/1.cd-hit_fatsa", help="Dir with cdhit fasta")
    parser.add_argument("--cg-file", default="output/CG_results/CG_ALL.txt", help="CG core gene table txt")
    parser.add_argument("--outbase", default="output/CG_results/all-strain-together", help="Output base directory")
    args = parser.parse_args()
    
    step2_dir = os.path.join(args.outbase, "2.result")
    step3_dir = os.path.join(args.outbase, "3.result")
    final_meg = os.path.join(args.outbase, "4.result", "all_core_gene.meg")
    final_log = os.path.join(args.outbase, "4.result", "4.log_order_len.txt")
    
    step2_get_fa(args.cdhit_dir, args.cg_file, step2_dir)
    step3_get_mega(step2_dir, step3_dir)
    step4_get_snp_mega(step3_dir, final_meg, final_log)
    logging.info(f"Done! Final mega file at {final_meg}")

File: genomevo/modules/bactcg_src/cmd/postprocess.go
package cmd

import (
"bufio"
"fmt"
"os"
"path/filepath"
"strings"
"github.com/spf13/cobra"
)

var (
inputCdhitRun   string
inputCgFileRun  string
baseOutDirRun   string
)

var getFaCmd = &cobra.Command{
Use:   "getfa",
Short: "Extract sequences from CD-HIT fasta to independent .fa files per family",
Run: func(cmd *cobra.Command, args []string) {
err := GetFaFileAll(inputCdhitRun, inputCgFileRun, baseOutDirRun)
if err != nil {
fmt.Println("Error:", err)
}
},
}

func init() {
rootCmd.AddCommand(getFaCmd)
getFaCmd.Flags().StringVarP(&inputCdhitRun, "cdhit", "c", "output/CG_results/1.cd-hit_output/1.cd-hit_fatsa", "Directory with cdhit FASTA")
getFaCmd.Flags().StringVarP(&inputCgFileRun, "cgfile", "g", "output/CG_results/CG_ALL.txt", "CG core gene table file")
getFaCmd.Flags().StringVarP(&baseOutDirRun, "outdir", "o", "output/CG_results/all-strain-together", "Base output directory")
}

func GetFaFileAll(inputCdhitDir, cgFile, baseOutDir string) error {
fmt.Println("Running Post-process Step 2: get_fa_file-all...")
outDir := filepath.Join(baseOutDir, "2.result")
os.MkdirAll(outDir, 0755)

seqDict := make(map[string]string)
files, err := os.ReadDir(inputCdhitDir)
if err != nil {
return fmt.Errorf("failed to read cdhit dir: %v", err)
}

for _, f := range files {
if !f.IsDir() && strings.HasSuffix(f.Name(), ".fasta") {
indexName := strings.TrimSuffix(f.Name(), ".fasta")
path := filepath.Join(inputCdhitDir, f.Name())
file, err := os.Open(path)
if err != nil {
continue
}
scanner := bufio.NewScanner(file)
var currentSeqName string
for scanner.Scan() {
line := scanner.Text()
if strings.HasPrefix(line, ">") {
currentSeqName = ">" + indexName + "---" + strings.TrimPrefix(line, ">")
seqDict[currentSeqName] = ""
} else if line != "" {
seqDict[currentSeqName] += strings.TrimSpace(line)
}
}
file.Close()
}
}

cgLineOutput := make([][]string, 0)
cgf, err := os.Open(cgFile)
if err != nil {
return fmt.Errorf("failed to open CG file: %v", err)
}
defer cgf.Close()

scanner := bufio.NewScanner(cgf)
if !scanner.Scan() {
return nil
}
firstLine := scanner.Text()
headers := strings.Split(firstLine, "\t")
var indexList []string
for _, h := range headers {
indexList = append(indexList, strings.TrimSpace(h)+"---")
}

for scanner.Scan() {
line := scanner.Text()
if line == "" {
continue
}
parts := strings.Split(line, "\t")
var outLine []string
for i := 0; i < len(parts) && i < len(indexList); i++ {
outLine = append(outLine, indexList[i]+strings.TrimSpace(parts[i]))
}
cgLineOutput = append(cgLineOutput, outLine)
}

for _, row := range cgLineOutput {
if len(row) == 0 {
continue
}
outFileName := strings.TrimPrefix(row[0], indexList[0])
outPath := filepath.Join(outDir, outFileName+".fa")
outf, err := os.Create(outPath)
if err != nil {
continue
}
writer := bufio.NewWriter(outf)
for _, el := range row {
key := ">" + el
if val, ok := seqDict[key]; ok {
writer.WriteString(key + "\n" + val + "\n")
}
}
writer.Flush()
outf.Close()
}
return nil
}

func GetSnpMegaAll(inputMegaDir, outputDir string) error {
	fmt.Println("Running Post-process Step 4: get_SNP_mega-all...")
	os.MkdirAll(outputDir, 0755)

	// Step 4a: For each .meg file, remove identical (non-informative) sites
	newMegaDir := filepath.Join(outputDir, "new_mega")
	os.MkdirAll(newMegaDir, 0755)

	files, err := os.ReadDir(inputMegaDir)
	if err != nil {
		return fmt.Errorf("failed to read mega dir: %v", err)
	}

	for _, f := range files {
		if !strings.HasSuffix(f.Name(), ".meg") {
			continue
		}

		inPath := filepath.Join(inputMegaDir, f.Name())
		outPath := filepath.Join(newMegaDir, "new-"+f.Name())

		if err := filterIdenticalSites(inPath, outPath); err != nil {
			fmt.Printf("Warning: SNP filter failed for %s: %v\n", f.Name(), err)
		}
	}

	// Step 4b: Concatenate all filtered MEGA files
	return concatMegaFiles(newMegaDir, outputDir)
}

// filterIdenticalSites removes positions where all sequences have identical amino acids.
func filterIdenticalSites(inFile, outFile string) error {
	in, err := os.Open(inFile)
	if err != nil {
		return err
	}
	defer in.Close()

	// Read sequences
	seqs := make(map[string]string)
	var seqOrder []string
	scanner := bufio.NewScanner(in)
	var currentId string
	inHeader := true

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "#") {
			if inHeader && !strings.HasPrefix(line, "#MEGA") && !strings.HasPrefix(line, "!") {
				// Sequence ID line
				currentId = strings.TrimPrefix(line, "#")
				seqOrder = append(seqOrder, currentId)
				seqs[currentId] = ""
				inHeader = false
			}
			continue
		}
		if strings.HasPrefix(line, ">") {
			currentId = strings.TrimPrefix(line, ">")
			seqOrder = append(seqOrder, currentId)
			seqs[currentId] = ""
			continue
		}
		if currentId != "" {
			seqs[currentId] += line
		}
	}

	if err := scanner.Err(); err != nil {
		return err
	}

	if len(seqOrder) == 0 || len(seqs[seqOrder[0]]) == 0 {
		// No sequences or empty - copy as-is
		return copyFile(inFile, outFile)
	}

	// Find positions where all sequences have identical AA
	seqLen := len(seqs[seqOrder[0]])
	keepPos := make([]bool, seqLen)
	for i := 0; i < seqLen; i++ {
		firstAA := seqs[seqOrder[0]][i]
		allSame := true
		for _, id := range seqOrder[1:] {
			if i >= len(seqs[id]) || seqs[id][i] != firstAA {
				allSame = false
				break
			}
		}
		keepPos[i] = !allSame // keep if NOT all identical
	}

	// Write filtered output
	out, err := os.Create(outFile)
	if err != nil {
		return err
	}
	defer out.Close()

	writer := bufio.NewWriter(out)
	writer.WriteString("#MEGA\n!Title CG;\n!Format DataType=Protein indel=-;\n\n")

	for _, id := range seqOrder {
		writer.WriteString(fmt.Sprintf("#%s\n", id))
		filtered := ""
		for i := 0; i < seqLen && i < len(seqs[id]); i++ {
			if keepPos[i] {
				filtered += string(seqs[id][i])
			}
		}
		writer.WriteString(filtered + "\n\n")
	}
	writer.Flush()
	return nil
}

// concatMegaFiles concatenates all filtered MEGA files into a single SNP alignment.
func concatMegaFiles(inputDir, outputDir string) error {
	files, err := os.ReadDir(inputDir)
	if err != nil {
		return fmt.Errorf("failed to read new_mega dir: %v", err)
	}

	allSeqs := make(map[string]string)
	var indexList []string
	var initLen int

	for _, f := range files {
		if !strings.HasSuffix(f.Name(), ".meg") {
			continue
		}

		path := filepath.Join(inputDir, f.Name())
		file, err := os.Open(path)
		if err != nil {
			continue
		}

		scanner := bufio.NewScanner(file)
		isFirst := (len(indexList) == 0)
		var currentId string

		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "" {
				continue
			}
			if strings.HasPrefix(line, "#") && !strings.HasPrefix(line, "#MEGA") && !strings.HasPrefix(line, "!") {
				currentId = strings.TrimPrefix(line, "#")
				if isFirst {
					indexList = append(indexList, currentId)
					allSeqs[currentId] = ""
				}
			} else if !strings.HasPrefix(line, "#") && !strings.HasPrefix(line, "!") {
				if _, ok := allSeqs[currentId]; ok {
					allSeqs[currentId] += line
				}
			}
		}
		file.Close()

		if isFirst && len(indexList) > 0 {
			initLen = len(allSeqs[indexList[0]])
		}
	}

	outPath := filepath.Join(outputDir, "all_core_gene.meg")
	outf, err := os.Create(outPath)
	if err != nil {
		return err
	}
	defer outf.Close()

	writer := bufio.NewWriter(outf)
	writer.WriteString("#MEGA\n!Title SNPMega;\n!Format DataType=Protein indel=-;\n\n")

	for _, id := range indexList {
		writer.WriteString(fmt.Sprintf("#%s\n", id))
		writer.WriteString(allSeqs[id] + "\n\n")
	}
	writer.Flush()

	fmt.Printf("Step 4 output to %s with concatenated length %d\n", outPath, initLen)
	return nil
}
// --- snpmega subcommand ---

var (
	snpmegaInputDir  string
	snpmegaOutputDir string
)

var snpmegaCmd = &cobra.Command{
	Use:   "snpmega",
	Short: "Concatenate MEGA alignments into SNP matrix",
	Long:  `Port of get_SNP_mega-all.py. Reads all .meg files in input dir, removes identical sites, concatenates sequences, and outputs a single all_core_gene.meg file.`,
	Run: func(cmd *cobra.Command, args []string) {
		if err := GetSnpMegaAll(snpmegaInputDir, snpmegaOutputDir); err != nil {
			fmt.Println("Error running snpmega:", err)
		}
	},
}

func init() {
	rootCmd.AddCommand(snpmegaCmd)
	snpmegaCmd.Flags().StringVarP(&snpmegaInputDir, "input", "i", "output/CG_results/all-strain-together/3.mega", "Directory containing .meg files")
	snpmegaCmd.Flags().StringVarP(&snpmegaOutputDir, "output", "o", "output/CG_results/all-strain-together/4.SNP_mega", "Output directory for SNP mega file")
}